Newsletters: add 120 (2020-10-21) (#478)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2020-10-21-newsletter.md

---
title: 'Bitcoin Optech Newsletter #120'
permalink: /en/newsletters/2020/10/21/
name: 2020-10-21-newsletter
slug: 2020-10-21-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter provides an overview of the new MuSig2 paper,
summarizes additional discussion about upfront fees in LN, and describes
a proposal to simplify management of LN payments.  Also included are our
regular sections with summaries of notable improvements to clients and
services, announcements of releases and release candidates, and changes
to popular Bitcoin infrastructure software.

## Action items

*None this week.*

## News

- MuSig2 FIXME:jnewbery

- **More LN upfront fees discussion:** after [last week's
  discussion][news119 upfront], Joost Jager asked developers on the
  Lightning-Dev mailing list to again [consider][jager hold fees] how to
  charge fees for attempting to route payments---fees that would have to
  be paid even if the attempt failed.  This could mitigate jamming

>   attacks that can prevent nodes from earning routing fees.

Reviewer: It seems to me that making network probing attacks (e.g. for tracking channel balances) more costly is also a central concern that is addressed by this proposal.

  HTLCs have been resolved and the channel is quiet---that improves
  safety but is more restrictive than desired.

    Russell's proposal this week was to instead only allow one party in
    a channel to propose HTLC changes.  This eliminates the risk of
    conflicting proposals but it increases the average network
    communication overhead by half a round trip (assuming each party
    proposes an equal number of HTLC changes).  Responsibility for
    proposing changes can be transferred from one party to the other as
    needed, allowing each party to propose channel updates at different
    times.

    As of this writing, the proposal received only a small amount of
    discussion from other LN implementation maintainers, so its future
    remains uncertain.

## Changes to services and client software

*In this monthly feature, we highlight interesting updates to Bitcoin
wallets and services.*

FIXME:bitschmidty

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

> FIXME:harding to update Tuesday

Reviewer 2: Bisq (if covered) just announced a release: https://github.com/bisq-network/bisq/releases/tag/v1.4.1

Author: I only cover the infrastructure projects listed in the Notable Code Changes section.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Hardware Wallet Interface (HWI)][hwi repo], [Bitcoin Improvement Proposals
(BIPs)][bips repo], and [Lightning BOLTs][bolts repo].*

- [Bitcoin Core #19953][] implements [schnorr signature
  verification][topic schnorr signatures] ([BIP340][]), [taproot][topic
  taproot] verification ([BIP341][]), and [tapscript][topic tapscript]
  verification ([BIP342][]).  The new soft fork rules are currently only

>   used in the private regression test mode (regtest); they are not

Reviewer 3: Suggest removing 'regression' here. That's the origin of the word 'regtest', since it was initially used with the bitcoinj comparison tool to find behavior changes (https://github.com/bitcoin/bitcoin/commit/0e4b31755534fac4ea6c20a60f719e3694252220), but we use it for all kinds of testing now.

  rules when an activation signal is detected.  The particular
  activation signal or signals to use are still being discussed in the
  [##taproot-activation][] IRC chatroom ([logs][##taproot-activation
  logs]).

    {% comment %}<!-- $ git diff 450d2b2371...865d2c37e2 | grep '^+' \
                        | grep -v '^+$' | grep -v '^+ */[/*]' | grep -v '^+ *\*' \
                        | grep -v '^+++' | wc -l
                      512
    -->{% endcomment %}

    The code [consists][sipa summary] of about [700 lines][sipa
    consensus] of consensus-related changes (500 excluding comments and
    whitespace) and [2,100 lines][sipa tests] of tests.  Over 30 people
    directly reviewed this PR and its [predecessor][Bitcoin Core
    #17977], and many others participated in developing and reviewing
    the underlying research, the BIPs, the related code in libsecp256k1,
    and other parts of the system.  There is still more work to do, but
    getting the code properly reviewed was arguably the most critical
    step---so we extend our most grateful appreciation to everyone who
    helped achieve this milestone.

> - [Bitcoin Core #19988][] refactors the logic Bitcoin Core uses to

Reviewer 3: s/refactors/redesigns/

'refactors' is too weak here. We usually use that to mean changes to the code that don't alter the logic or functionality. This PR was a complete redesign of the way we select which peers to download transactions from.

  request transactions from its peers.  The major changes described by
  the PR description include: Bitcoin Core will now more strongly prefer
  requesting transactions from *outbound peers*---peers where the
  connection was opened by the local node; Bitcoin Core will now request
  an unlimited number of transactions from a peer rather than stopping
  at 100 (though slow responders won't be preferred); and Bitcoin Core
  won't keep track of as many pending transaction requests as before,
  the previous limit being seen as excessive.  The change should also
  make it easier to test that the transaction request logic is working

>   as expected, with multiple tests for that included in the merged PR.

Reviewer 3: 'multiple tests' is selling this PR short:

- an easy-to-review naive implementation and fuzz tests demonstrating the equivalence between the naive and efficient implementations
- almost 100% unit test coverage with
- multiple functional tests for testing the high-level functionality.

This is probably now the best tested code in Bitcoin Core!

>   This may help reduce the chance that a flaw can be exploited against

Reviewer 2: Does "This" refer to the easier testing/multiple tests, or to the PR?

  users of time-sensitive transactions (such as in LN and many other

>   contract protocols).

Reviewer 3: Other things that might be worth mentioning:

- better specified bounds on memory and CPU complexity
- cleaner interfaces and better code structure
- very thorough testing (see above) and review (over 600 review comments!)

Not sure how much of that we need to include for readers.

Author: We usually don't cover testing or review at all, which has always been unfair to people who put significant work into that.  I cover that a bit for taproot this week because some people considering adopting it will want to know how much review it received, but in general I try to focus on how the PR will affect users of the software.

- [Bitcoin Core #19077][] wallet: Add sqlite as an alternative wallet database and use it for new descriptor wallets FIXME:dongcarl

- [Bitcoin Core #19770][] RPC: getpeerinfo: Deprecate "whitelisted" field (replaced by "permissions") FIXME:bitschmidty

> - [Bitcoin Core #17428][] periodically writes a file with the network

Reviewer 3: Actually it only writes the anchor connection on shutdown, but I think leading with the fact that we write a file is focusing on the implementation details rather than the big picture. We should start with 'persists up to two connections across restarts' or similar.

Author: I'll have to check the code, but wouldn't writing the file only on shutdown defeat the purpose of dealing with *crashes*?

Author: Hmm.  Looks like I misunderstood the PR.

  addresses of the node's two outbound block-relay-only peers.  The next
  time the node starts, it reads this file and attempts to reconnect to

>   those same two peers.  This prevents an attacker with the ability to

Reviewer 3: This isn't true. We only write the anchor connections file on a clean shutdown, so if an attacker can make us crash, then we won't persist connections.

  indicate how long until LND stops making additional attempts to send a
  payment.  The default remains 60 seconds.

- [Libsecp256k1 #830][] enables GLV endomorphism, which allows verifying
  signatures with up to 33% fewer operations (see a comment to the PR
  with actual [performance test results][o'beirne bitcoinperf]).  The PR
  also removes the slower code for operating without endomorphism as
  there's no reason to use it now that the patent on the GLV technique
  has expired (see [Newsletter #117][news117 glv]).  This PR also
  implements an independent `memcmp` (memory compare) function to avoid
  the problems with GCC's internal version of that function (see
  [another section][news117 memcmp] in Newsletter #117).  These changes
  were subsequently [pulled][Bitcoin Core #20147] into Bitcoin Core.

<!-- FIXME: harding to update topic references Tuesday -->
{% include references.md %}
{% include linkers/issues.md issues="19953,17977,19988,19077,19770,17428,4688,830,20147" %}
[sipa summary]: https://github.com/bitcoin/bitcoin/pull/19953#issuecomment-691815830
[sipa consensus]: https://github.com/sipa/bitcoin/compare/450d2b2371...865d2c37e2

> [sipa tests]: https://github.com/sipa/bitcoin/compare/206fb180ec...0e2a5e448f426219a6464b9aaadcc715534114e6

Reviewer 3: Can these point to bitcoin/bitcoin hashes now that the commits are merged into master?

---
This week's newsletter provides an overview of the new MuSig2 paper,
summarizes additional discussion about upfront fees in LN, and describes
a proposal to simplify management of LN payments.  Also included are our
regular sections with summaries of notable improvements to clients and
services, announcements of releases and release candidates, and changes
to popular Bitcoin infrastructure software.

## Action items

*None this week.*

## News

- **MuSig2 paper published:** Jonas Nick, Tim Ruffing, and Yannick Seurin
  published [the MuSig2 paper][musig 2 paper] describing a new variant of the
  [MuSig][topic musig] signature scheme with a two round signing protocol.

    MuSig is a signature scheme that allows multiple signers to create an
    aggregate public key derived from their individual private keys, and then
    collaborate to create a single valid signature for that public key. The
    aggregate public key and signature are indistinguishable from any other
    [schnorr][topic schnorr signatures] public key and signature.

    The original version of MuSig required a three round signing protocol:
    first, the co-signers exchange commitments to nonce values, then they

>     exchange the nonce values themselves, and finally they exchange the

Reviewer 4: ```suggestion
    exchange the nonce values themselves, and finally, they exchange the
```

Reviewer 3: done

    partial signatures. This three round protocol is to avoid nonce reuse and

>     related attacks, which could allow a dishonest signer to discover the private

Reviewer 4: nit: wordy

```suggestion
    related attacks, allowing a dishonest signer to discover the private
```

Reviewer 3: I've reworked this whole sentence after Tim's review comment above.

>     key of one of their co-signers. Using deterministic nonces, which is very

Reviewer 5: Hm, having three rounds is not related to nonce reuse.

Suggestion:
This three round protocol is to avoid that an attacker interacting with an honest signer in multiple concurrent signing sessions can obtain a signature on a message that the honest signer did not want to sign. 

(This is really about obtaining a single bad signature (a forgery) but not discover the private key.)

Reviewer 3: Thanks! Fixed.

    common practice in single-signer schemes, is unsafe for multi-signer schemes,
    as described in the [original MuSig blog post][musig blog post unsafe
    deterministic nonce].  Precomputing the nonces in advance and exchanging them

>     at key setup time is also unsafe, as described in a [blog post by Jonas

Reviewer 4: nit: unsafe used in previous sentence
```suggestion
    at key setup time is also dangerous, as described in a [blog post by Jonas
```

Reviewer: repeating `unsafe` hear seems fine to me, since it's essentially a technical term.

>     Nick][unsafe nonce sharing]. It _is_ however safe to precompute the nonces and

Reviewer 4: ```suggestion
    Nick][unsafe nonce sharing]. It _is_, however, safe to precompute the nonces and
```

    exchange the nonce commitments early, moving one of the three rounds to the key
    setup stage.

>     Removing the requirement to exchange nonce commmittments has been an active

Reviewer 4: ```suggestion
    Removing the requirement to exchange nonce commitments has been an active
```

Reviewer 3: ttthannks!

    area of research, and last month the [MuSig-DN paper][musig-dn] was published, demonstrating
    how nonce commitment exchange could be removed by generating the nonce
    deterministically from the signers' public keys and the message, and providing a
    non-interactive zero-knowledge proof that the nonce was generated
    deterministically along with the nonce. This removed the requirement of
    exchanging nonce commitments at the cost of a more expensive signing operation.

>     The MuSig2 scheme achieves a two round signing protocol without the need for a

Reviewer 5: nit: Maybe add "simple" two round signing protocol.

Reviewer 3: done

    zero-knowledge proof. What's more, the first round (nonce exchange) can be done
    at key setup time, allowing two different variants:

    - interactive setup (key setup and nonce exchange), and non-interactive signing
    - non-interactive setup (address computed from public keys) and interactive
      signing (nonce exchange followed partial signature exchange)

    The non-interactive signing variant could be particularly useful for cold storage

>     schemes and offline signers.

Reviewer 5: Another prominent potential use case is LN, where you could exchange nonces already when you setup a channel.

Reviewer 3: Added. Thanks!

    MuSig2 also makes it possible to privately nest participants (e.g. if Alice
    and Bob create an aggregate public key using MuSig2, Bob's contribution
    could actually be an aggregate of Bob and Charlie's keys, without revealing
    anything to Alice). This could be particularly useful when [using multi

>     signer schemes in offchain contract protocols like LN][composable musig in ln].

Reviewer 5: This is currently not true. We are planning to work on this but we didn't get this ready for the release of this paper. In particular, we don't have a security proof so far. So if we want to mention this, it should be made clear that this is only future work. (It's maybe a good idea to mention this to clarify exactly this. There is some misunderstanding around this because we were talking about this earlier but didn't include it for now.)

Reviewer 3: Thanks for the correction! I'm just going to leave this out entirely, since I think including it is likely to just add confusion.

- **More LN upfront fees discussion:** after [last week's
  discussion][news119 upfront], Joost Jager asked developers on the
  Lightning-Dev mailing list to again [consider][jager hold fees] how to
  charge fees for attempting to route payments---fees that would have to
  be paid even if the attempt failed.  This could mitigate jamming
  attacks that can prevent nodes from earning routing fees.
  Unfortunately, developers have not yet found a satisfactory way to
  trustlessly charge upfront fees, so recent discussion has focused on
  trust-based methods that may involve such small amounts of fee that
  users will find them acceptable.

    Unfortunately, none of the methods discussed this week seemed to
    meet with widespread acceptance.  Several developers expressed
    concern that methods could be abused to penalize small honest nodes.
    Other developers noted that the alternative to upfront payment is

>     likely an increased reliance on reputation---which likely

Reviewer 4: repetitive 'likely'

```suggestion
    an increased reliance on reputation---which likely
```

    disproportionately benefits larger nodes.  We're sure developers will
    continue working on this important issue and we'll report on any
    notable progress in future newsletters.

- **Simplified HTLC negotiation:** Rusty Russell [posted][russell
  simplified] to the Lightning-Dev mailing list about simplifying how
  payments are made and resolved in LN channels.  Currently, both sides
  of a channel can each propose relaying or settling a payment, which
  requires creating or removing an [HTLC][topic htlc] from the offchain
  commitment transaction.  Sometimes both sides propose a change at the
  same time and so there are conflicting commitment transactions.  These
  potential conflicts may get more complicated if dynamic changes to the
  commitment transaction format and settings are allowed (see
  [Newsletter #108][news108 upgrade commitment]).  One proposal to
  eliminate that complication is to only allow dynamic changes when all
  HTLCs have been resolved and the channel is quiet---that improves
  safety but is more restrictive than desired.

>     Russell's proposal this week was to instead only allow one party in

Reviewer 4: nit: split infinitive

```suggestion
    Russell's proposal this week was to allow only one party in
```

    a channel to propose HTLC changes.  This eliminates the risk of

>     conflicting proposals but it increases the average network

Reviewer 4: ```suggestion
    conflicting proposals, but it increases the average network
```

    whitespace) and [2,100 lines][sipa tests] of tests.  Over 30 people
    directly reviewed this PR and its [predecessor][Bitcoin Core
    #17977], and many others participated in developing and reviewing
    the underlying research, the BIPs, the related code in libsecp256k1,
    and other parts of the system.  There is still more work to do, but
    getting the code properly reviewed was arguably the most critical
    step---so we extend our most grateful appreciation to everyone who
    helped achieve this milestone.

- [Bitcoin Core #19988][] redesigns the logic Bitcoin Core uses to
  request transactions from its peers.  The major changes described by
  the PR description include: Bitcoin Core will now more strongly prefer
  requesting transactions from *outbound peers*---peers where the
  connection was opened by the local node; Bitcoin Core will now request
  an unlimited number of transactions from a peer rather than stopping
  at 100 (though slow responders won't be preferred); and Bitcoin Core
  won't keep track of as many pending transaction requests as before,
  the previous limit being seen as excessive.  The change also greatly

>   improves the readability of the code and ability to test that the

Reviewer 4: ```suggestion
  improves the readability of the code and the ability to test that the
```

  transaction request logic is working as expected.  This testing may
  help reduce the chance that a flaw can be exploited against
  users of time-sensitive transactions (such as in LN and many other
  contract protocols).

- [Bitcoin Core #19077][] wallet: Add sqlite as an alternative wallet database and use it for new descriptor wallets FIXME:dongcarl

- [Bitcoin Core #19770][] deprecates the `whitelisted` field returned from the
  `getpeerinfo` RPC, which previously had its scope expanded by more granular permissioning
  (see [Newsletter #60][news60 permissions]). `whitelisted` is set to be
  deprecated in v0.21 and is planned to be removed in v0.22.

- [Bitcoin Core #17428][] writes a file at shutdown with the network
  addresses of the node's two outbound block-relay-only peers.  The next
  time the node starts, it reads this file and attempts to reconnect to
  those same two peers.  This prevents an attacker from using node
  restarts to trigger a complete change in peers,
  which would be something they could use as part of an [eclipse

>   attack][topic eclipse attacks] that could potentially be used to trick

Reviewer 4: nit: cutting some words
```suggestion
  attack][topic eclipse attacks] that could potentially trick
```

  users of time-sensitive transactions (such as in LN and many other
  contract protocols).

- [Bitcoin Core #19077][] adds a new SQLite database backend for wallet records.
  Unlike the existing Berkeley DB database backend, SQLite is designed and
  tested to be resilient against crash and power loss corruption. SQLite 3 also
  has a backward compatibility guarantee, the lack of which in Berkeley DB is
  the reason why Bitcoin Core has to target Berkeley DB version 4.8
  specifically. This backward compatibility guarantee also means that Bitcoin
  Core can benefit from the bug fixes and system compatibility improvements in

>   future releases of SQLite 3. In the upcoming v0.21 release, legacy wallets

Reviewer 6: Not sure that "This backward compatibility guarantee also means that Bitcoin Core can benefit from the bug fixes and system compatibility improvements in future releases of SQLite 3" is entirely relevant to readers. Reviewers: feel free to remove!

Reviewer 3: I agree that this is getting into the weeds a bit. Let's cut it.

Reviewer 3: I'd also suggest removing the backwards-compatibility sentence before. bitcoind can be built with newer versions of bdb, but by default we don't (for reasons that don't entirely make sense to me).
